### PR TITLE
8260009: InstanceKlass::has_as_permitted_subclass() fails if subclass was redefined

### PR DESCRIPTION
**Pull request by hseigel**
Please review this fix for JDK-8260009.  The fix removes the "if (k2 == k)" check for a valid permitted subclass because the check does not work if "k" is being redefined.  It fails because 'k' is a temporary InstanceKlass created when loading the redefined class.

Instead, the code relies on a name check to determine if 'k' is a permitted subclass.

The fix was tested with Mach5 tiers 1-2 on Linux, Mac OS, and Windows, tiers 3-5 on Linux x64, and JCK Lang and VM tests.

Thanks, Harold

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260009](https://bugs.openjdk.java.net/browse/JDK-8260009): InstanceKlass::has_as_permitted_subclass() fails if subclass was redefined

### Reviewers
 * [Lois Foltan](https://openjdk.java.net/census#lfoltan) (@lfoltan - **Reviewer**) ⚠️ Review applies to ca70228f587e70af08356448d79d7e142ed033cd
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to ca70228f587e70af08356448d79d7e142ed033cd
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to ca70228f587e70af08356448d79d7e142ed033cd

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2184/head:pull/2184`
`$ git checkout pull/2184`

### CODE REVIEW

**Comment by dcubed-ojdk** [comment]
@sspitsyn - This issue might interest you.

**Review by lfoltan — APPROVED** [review]
Looks good Harold!
Lois

**Review by sspitsyn — APPROVED** [review]
Hi Harold,
The fix looks good to me.
Thanks,
Serguei

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 1]
```diff
@@ -250,23 +250,15 @@ bool InstanceKlass::has_as_permitted_subclass(const InstanceKlass* k) const {
     return false;
   }
 
-  // Check for a resolved cp entry, else fall back to a name check.
+  // Do not check for a resolved cp entry, because that check can fail if
+  // the class is being redefined.  Just do a name check.
   // We don't want to resolve any class other than the one being checked.
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/instanceKlass.cpp:255]: You don't resolve any classes at all now so this comment can go.

**Review by dholmes-ora — APPROVED** [review]
Hi Harold,

This makes me wonder whether we can have a similar issue with nestmate checking?

Just curious: is there any way to determine that a Klass k2 is actually a redefinition of Klass k1? I

One nit below, otherwise this seems fine.

Thanks,
David

**Comment by hseigel** [comment]
Hi David,

I agree that this is not an issue for nestmates.  Class redefinition loads the redefined class, and class loading calls verify_member_access(), which does a nestmate check for private access.  But, class loading only calls verify_member_access() for accessing non-private methods, when doing method override checking in check_final_method_override().

Thanks, Harold
________________________________

File: src/hotspot/share/oops/instanceKlass.cpp (hunk 1 above)
Comment (hseigel) [reply src/hotspot/share/oops/instanceKlass.cpp]: comment was deleted.

**Comment by hseigel** [comment]
Thanks Lois, Serguei, and David for the reviews!

/integrate

**Pull request closed without merging**